filestorage.go

```diff
@@ -149,33 +151,46 @@ func dataDir() string {
 	return filepath.Join(baseDir, "certmagic")
 }
 
-// TryLock attempts to get a lock for name, otherwise it returns
-// a Waiter value to wait until the other process is finished.
-func (fs FileStorage) TryLock(key string) (Waiter, error) {
+// Lock obtains a lock named by the given key. It blocks
+// until the lock can be obtained or an error is returned.
+func (fs FileStorage) Lock(key string) error {
+	// can't defer the unlock because we might have
+	// to Wait() for a while before returning, so we're
+	// careful to unlock at all the right places
 	fileStorageNameLocksMu.Lock()
-	defer fileStorageNameLocksMu.Unlock()
+
+	if fs.fileStorageNameLocks == nil {
+		fs.fileStorageNameLocks = make(map[string]*fileStorageWaiter)
+	}
 
 	// see if lock already exists within this process - allows
 	// for faster unlocking since we don't have to poll the disk
-	fw, ok := fileStorageNameLocks[key]
+	fw, ok := fs.fileStorageNameLocks[key]
 	if ok {
 		// lock already created within process, let caller wait on it
-		return fw, nil
+		fileStorageNameLocksMu.Unlock()
+		fw.Wait()
+		return nil
 	}
 
 	// attempt to persist lock to disk by creating lock file
 
 	// parent dir must exist
 	lockDir := fs.lockDir()
 	if err := os.MkdirAll(lockDir, 0700); err != nil {
```

Comment:
This mkdir is outside the loop below, so if another process unlocks after this, they might also remove the directory and we will fail to create the lock file.

I reworked the lock acquisition in https://github.com/DisposaBoy/certmagic/commit/b5e59fc352bdf4e6318452d6a315f92ef8827cb3 so I could continue testing.